// @mattrglobal/node-bbs-signatures/lib/index.js
"use strict";
/*
 * Copyright 2020 - MATTR Limited
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
var bls12381_1 = require("./bls12381");
Object.defineProperty(exports, "generateBls12381G1KeyPair", { enumerable: true, get: function () { return bls12381_1.generateBls12381G1KeyPair; } });
Object.defineProperty(exports, "generateBls12381G2KeyPair", { enumerable: true, get: function () { return bls12381_1.generateBls12381G2KeyPair; } });
Object.defineProperty(exports, "generateBlindedBls12381G1KeyPair", { enumerable: true, get: function () { return bls12381_1.generateBlindedBls12381G1KeyPair; } });
Object.defineProperty(exports, "generateBlindedBls12381G2KeyPair", { enumerable: true, get: function () { return bls12381_1.generateBlindedBls12381G2KeyPair; } });
var bls12381toBbs_1 = require("./bls12381toBbs");
Object.defineProperty(exports, "bls12381toBbs", { enumerable: true, get: function () { return bls12381toBbs_1.bls12381toBbs; } });
var bbsSignature_1 = require("./bbsSignature");
Object.defineProperty(exports, "BBS_SIGNATURE_LENGTH", { enumerable: true, get: function () { return bbsSignature_1.BBS_SIGNATURE_LENGTH; } });
Object.defineProperty(exports, "sign", { enumerable: true, get: function () { return bbsSignature_1.sign; } });
Object.defineProperty(exports, "blsSign", { enumerable: true, get: function () { return bbsSignature_1.blsSign; } });
Object.defineProperty(exports, "verify", { enumerable: true, get: function () { return bbsSignature_1.verify; } });
Object.defineProperty(exports, "blsVerify", { enumerable: true, get: function () { return bbsSignature_1.blsVerify; } });
Object.defineProperty(exports, "createProof", { enumerable: true, get: function () { return bbsSignature_1.createProof; } });
Object.defineProperty(exports, "blsCreateProof", { enumerable: true, get: function () { return bbsSignature_1.blsCreateProof; } });
Object.defineProperty(exports, "verifyProof", { enumerable: true, get: function () { return bbsSignature_1.verifyProof; } });
Object.defineProperty(exports, "blsVerifyProof", { enumerable: true, get: function () { return bbsSignature_1.blsVerifyProof; } });
__exportStar(require("./types"), exports);
